File: C++/GraphsCPP/clone_Graph.cpp
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == NULL)
            return NULL;
        
        queue<Node*> q;
        q.push(node);        
        vector<Node*> res(101, NULL);

        res[node->val] = new Node(node->val);
        
        while(!q.empty()){
            int size = q.size();
            
            for(int i = 0; i < size; i++){
                auto current = q.front();
                q.pop();
                
                for(auto next: current->neighbors){
                    if(!res[next->val]){
                        res[next->val] = new Node(next->val);
                        q.push(next);
                    }
                    
                    res[current->val]->neighbors.push_back(res[next->val]);
                }
            }
        }
        return res[node->val];
    }
};

File: JAVA/GraphsJava/CloneGraph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null)
        return null;
        Queue<Node> que = new LinkedList<>();
        que.add(node);
        HashMap<Node, Node> map = new HashMap<>();
        Node nnode = new Node(node.val, new ArrayList<>());
        map.put(node, nnode);
        while (!que.isEmpty()) {
            Node temp = que.poll();
            for (Node i: temp.neighbors) {
                if (map.containsKey(i) == false) {
                    que.add(i);
                    map.put(i, new Node(i.val, new ArrayList<>()));
                }
                map.get(temp).neighbors.add(map.get(i));
            }
        }
    return nnode;
    }
}
